File: flight_controller/components/pid.py
import time


class PID:

    def __init__(self, kp=2.0, ki=0.0, kd=0.0, set_point=0.0, sample_time=0.05, output_max=500.0, output_min=-500.0):
        self._kp, self._ki, self._kd = kp, ki, kd
        self._set_point = set_point
        self._output_max = output_max
        self._output_min = output_min
        self._sample_time = sample_time
        self._proportional = 0
        self._integral = 0
        self._derivative = 0
        self._last_time = time.time()
        self._last_output = None
        self._last_error = None

    def __call__(self, current_value):
        current_time = time.time()
        dt = current_time - self._last_time if current_time - self._last_time else 1e-16
        if dt < self._sample_time and self._last_output is not None:
            return self._last_output
        error = self._set_point - current_value
        d_error = error - self._last_error if self._last_error is not None else 0
        self._proportional = self._kp * error
        self._integral += self._ki * error * dt
        self._integral = max(min(self._integral, self._output_max), self._output_min)
        self._derivative = -self._kd * d_error / dt
        output = self._proportional + self._integral + self._derivative
        output = max(min(output, self._output_max), self._output_min)
        self._last_error = error
        self._last_output = output
        self._last_time = current_time
        return output

    def set_point(self, set_point):
        self._set_point = set_point

    def set_components(self, kp=0.0, ki=0.0, kd=0.0):
        self._kp, self._ki, self._kd = kp, ki, kd

    def set_limits(self, output_max=1.0, output_min=-1.0):
        self._output_max, self._output_min = output_max, output_min

    def set_sample_time(self, sample_time=0.01):
        self._sample_time = sample_time

    def reset(self):
        self._proportional = 0
        self._integral = 0
        self._derivative = 0
        self._last_time = time.time()
        self._last_output = None
        self._last_error = None

File: flight_controller/libraries/FlightController/PID.cpp
#include <Arduino.h>
#include "PID.h"

PID::PID() {
	SetConstants(1.0, 0.0, 0.0);
	SetDesiredValue(0.0);
	SetSampleTime(50000ul);
	SetOutputLimits(500.0, -500.0);
	ResetValues();
}

PID::PID(float kp, float kd, float ki, float desiredValue, unsigned long sampleTime, float maxOutput, float minOutput) {
	SetConstants(kp, kd, ki);
	SetDesiredValue(desiredValue);
	SetSampleTime(sampleTime);
	SetOutputLimits(maxOutput, minOutput);
	ResetValues();
}

PID::PID(float kp, float kd, float ki) {
	SetConstants(kp, kd, ki);
	SetDesiredValue(0.0);
	SetSampleTime(50000ul);
	SetOutputLimits(500.0, -500.0);
	ResetValues();
}

PID::PID(float kp, float kd, float ki, float desiredValue, float sampleTime, float maxOutput, float minOutput) {
	SetConstants(kp, kd, ki);
	SetDesiredValue(desiredValue);
	SetSampleTime(sampleTime);
	SetOutputLimits(maxOutput, minOutput);
}

void PID::SetSampleTime(float sampleTime) {
	float microsecondsInSecond = 1000000.0;
	_sampleTime = sampleTime * microsecondsInSecond;
}

void PID::ResetValues() {
	_integral = 0.0;
	_lastError = 0.0;
	_lastOutput = 0.0;
	_lastUpdateTime = 0;
	_outputCalculated = false;
	_derivativeCalculated = false;
}

void PID::SetConstants(float kp, float kd, float ki) {
	_kp = kp;
	_kd = kd;
	_ki = ki;
}

float PID::ClampValue(float value, float min, float max) {
	if (value > max) {
		return max;
	}
	if (value < min) {
		return min;
	}
	return value;
}

float PID::Update(float actualValue) {
	float microsecondsInSecond = 1000000.0;
	float output, error, derror, pid_p, pid_d, pid_i;
	unsigned long actualTime, elapsedTime;
	float elapsedTimeInSeconds;
	actualTime = micros();
	elapsedTime = actualTime - _lastUpdateTime;
	if (elapsedTime < _sampleTime && _outputCalculated) {
		return _lastOutput;
	}
	elapsedTimeInSeconds = elapsedTime / microsecondsInSecond;
	error = _desiredValue - actualValue;
	derror = error - _lastError;
	pid_p = _kp * error;
	if (_derivativeCalculated) {
		pid_d = _kd * (derror / elapsedTimeInSeconds);
	} else {
		pid_d = 0;
		_derivativeCalculated = true;
	}
	_integral += _ki * (error * elapsedTimeInSeconds);
	_integral = ClampValue(_integral, _minOutput, _maxOutput);
	pid_i = _integral;
	output = pid_p + pid_d + pid_i;
	output = ClampValue(output, _minOutput, _maxOutput);
	_outputCalculated = true;
	_lastUpdateTime = actualTime;
	_lastError = error;
	_lastOutput = output;
	return output;
}

void PID::SetDesiredValue(float desiredValue) {
	_desiredValue = desiredValue;
	ResetValues();
}

void PID::SetSampleTime(unsigned long sampleTime) {
	_sampleTime = sampleTime;
}

void PID::SetOutputLimits(float maxOutput, float minOutput) {
	_maxOutput = maxOutput;
	_minOutput = minOutput;
}